### PR TITLE
DRYD-54: Add RecordHeader component and dock on scroll.

### PR DESCRIPTION
**Pull request by ynejati**

### CODE REVIEW

**Comment by ynejati** [comment]
There are two things that I'm not sure I'm happy with, first, the use of media queries to determine the width of the RecordHeader as the viewport resizes, second, setting the minimum height of the TitleBar so that location of the RecordHeader can be statically set (top: 70;). I used the media queries because I could not figure out how to have the header maintain the same width as the RecordEditor on resize. I set the min-height of the TitleBar so that the needed position for the header doesn't need to vary between new records and existing records as the content of the TitleBar changes, and thus it's height.  Does anyone have any suggestions for improvement on these items? Thoughts?

**Comment by ray-lee** [comment]
I agree with your reservations. It does mostly work though, and I think it can be refined to be good enough. But let me think about it.

**Comment by ynejati** [comment]
Glad you agree. At the moment I can't think of a better way unless we restructure things a bit. We could just as easily add more media queries to solve the width issue, or just go with 100% width and call it, but that would take up too much space. For the position of one fixed element to the other, I'm not sure if there's any any way to do that dynamically using CSS. Also, I'm making the naive assumption that the title bar will never wrap.

**Comment by ray-lee** [comment]
Yousuf, I'm going to send you a pull request with css changes that remove the media queries, and make the docked bar look the way I'd like. It jumps around a bit vertically when it docks, but I didn't try to address that. I'd like it to be smooth.

To deal with the title bar height, let's add a callback prop to TitleBar: onHeightChanged(height). And let's add a prop to RecordHeader: dockTop. Then RecordPage can hook them up.

**Comment by ynejati** [comment]
Sounds like a plan. I will get to it.

**Comment by ynejati** [comment]
Ready for review now. The RecordPage now has a state property that is set by the titlebar height change handler which then gets passed down to the RecordHeader as a prop to help determine where it should dock using inline styles. Inline styles are also used to ensure that the editor form doesn't jump upward when the RecordHeader is docked and thus a much smoother experience on scroll. Let me know what you think. By the way, I like the way you faded the right edge of the header into the editor by manipulating the box shadow. You're right, it is much cleaner.

File: src/components/pages/RecordPage.jsx [hunk 1]
```diff
@@ -139,8 +144,19 @@ export default class RecordPage extends Component {
     });
   }
 
+  // Handle positioning of RecordHeader on dock depending on TitleBar height.
+  handleTitleBarHeightChanged(height) {
+    this.setState({
+      headerDockPosition: height - 10,
```
Comment (ray-lee) [inline src/components/pages/RecordPage.jsx]: I don't like having 10 hardcoded here.

File: src/components/record/RecordHeader.jsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    this.domNode = ref;
+  }
+
+  handleScroll() {
+    const node = this.domNode;
+
+    if (!node) return;
+
+    if (this.state.docked) {
+      if (window.scrollY < 136) {
```
Comment (ray-lee) [inline src/components/record/RecordHeader.jsx]: I don't like having 136 hardcoded.

File: src/components/record/RecordHeader.jsx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        this.setState({
+          docked: false,
+        });
+      }
+    } else if (window.scrollY >= 136) {
```
Comment (ray-lee) [inline src/components/record/RecordHeader.jsx]: Ditto.

File: src/components/record/RecordHeader.jsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      onCloneButtonClick,
+      onDeleteButtonClick,
+    } = this.props;
+
+    const {
+      docked,
+    } = this.state;
+
+    const className = classNames(docked ? styles.docked : '');
+    const containerInlineStyle = docked ? { minHeight: '42px' } : {};
```
Comment (ray-lee) [inline src/components/record/RecordHeader.jsx]: I don't want to hardcode 42.

File: src/components/sections/TitleBar.jsx [hunk 5]
```diff
@@ -53,6 +58,10 @@ export default class TitleBar extends Component {
       this.setState({
         docked: true,
       });
+
+      if (onHeightChanged) {
+        onHeightChanged(node.offsetHeight);
```
Comment (ray-lee) [inline src/components/sections/TitleBar.jsx]: Hmm, we're not actually notifying that the height has changed here. We're notifying that the title bar has docked, and passing the height at the time of docking. Which I think is fine -- but let's rename this prop to onDocked.

Comment (ynejati) [reply src/components/sections/TitleBar.jsx]: Hey Ray, thanks for reviewing. In general, I will avoid hardcoding anything in the future. I look forward to reviewing your changes and let me know if there's anything I can do to help. Thanks again!

**Comment by ray-lee** [comment]
Thanks! People will love this.

**Pull request merged**